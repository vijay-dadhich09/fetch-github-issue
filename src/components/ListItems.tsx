import React from 'react';
import {View, Text, FlatList, StyleSheet, TouchableOpacity} from 'react-native';

interface ListProps {
	urls: any[], 
	issues: any,
	loading: boolean,
	hasErrors: boolean,
	loadMore: Function,
	currentPage: number,
	totalCount: number
}
const renderItem = ({item}) => {
	return (
		<View key={item.id} style={styles.itemContainer}>
			<Text style={styles.text}>{item.title}</Text>
		</View>
	)
}

const keyExtractor = (item: any) => item.id;

const renderSeparator = () => {
	return <View style={styles.separatorStyle} />
}
const renderPaggination = (urls: any[], loadMore: Function, currentPage: number, totalCount: number) => {
	const totalPages = Math.ceil(totalCount / 30);
	const result  = urls.map((item:any, index: number) => {
		if (item.showPage) {
			return (
				<Text key={index} style={styles.pageText}>{`Page ${currentPage} of ${totalPages}`}</Text>
			)
		}
		return (
			<>
				<TouchableOpacity
					key={index}
					style={styles.navBar}
					onPress={() => loadMore(item.url, item.page)}
				>
					<Text style={styles.navText}>{item.title}</Text>
				</TouchableOpacity>
			</>
		)
	})
	return result;
}

function ListItems (props: ListProps) {
	const { loading, hasErrors, issues, urls, currentPage, loadMore, totalCount } = props;
	const { items } = issues && issues[currentPage] || {};
	// console.log('data-loading: ', loading);
  return (
		<View style={styles.container}>
			<Text style={styles.titleText}>List of Issues</Text>
			{urls && urls.length>0 && <View style={styles.navBarContainer}>
				{renderPaggination(urls, loadMore, currentPage, totalCount)}
			</View>}
			<FlatList
          contentContainerStyle={styles.ListContainer}
          data={items} 
          renderItem={renderItem}
          keyExtractor={keyExtractor}
					ItemSeparatorComponent={renderSeparator}
        />
		</View>
	)
}

export default React.memo(ListItems);
const styles = StyleSheet.create({
	container: {
		width: '100%',
		flex: 1,
		padding: 10,
	},
	ListContainer: {
		marginTop: 10,
		paddingBottom: 10
	},
	separatorStyle: {
		height: 5
	},
	itemContainer: {
		padding: 10,
		backgroundColor: '#5DADE2',
		minHeight: 60,
		justifyContent: 'center'
	},
	text: {
		color: "#09182b",
		fontSize: 15,
		fontWeight: "800",
	},
	navBarContainer: {
		marginTop: 10,
		marginBottom: 10,
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center'
	},
	navBarInnerContainer: {
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center'
	},
	navBar: {
		padding: 10,
		backgroundColor: '#C0392B',
	},
	navText: {
		color: '#FDEDEC',
		fontSize: 15,
		fontWeight: "800",
		textAlign: 'center'
	},
	titleText: {
		color: '#273746',
		fontSize: 20,
		fontWeight: "800",
		textAlign: 'center'
	},
	pageText: {
		color: '#273746',
		fontSize: 15,
		fontWeight: "800",
		textAlign: 'center'
	}
})