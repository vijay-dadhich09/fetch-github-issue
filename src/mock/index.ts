export const ISSUES_DATA = {
    urls: [
      {
        url: 'https://api.github.com/search/issues?q=repo%3Amicrosoft%2FCBL-Mariner&page=2',
        title: 'next',
        page: '2',
        showPage: false
      },
      {
        showPage: true
      },
      {
        url: ' https://api.github.com/search/issues?q=repo%3Amicrosoft%2FCBL-Mariner&page=27',
        title: 'last',
        page: '27',
        showPage: false
      }
    ], 
    issues: {
      '1': {
        total_count: 1,
        incomplete_results: false,
        items: [
          {
            id: 842912478,
            title: 'Automatic update of the `kubernetes` packages.',
          },
          {
            id: 842668222,
            number: 805,
            title: 'diskutils: add more robust handling of disk/partition operations',
          },
          {
            id: 842369440, 
            title: 'Change dropdown menus to accessible input fields',
          },
        ]
      }
    },
    loading: false,
    hasErrors: false,
    loadMore: () => {},
    currentPage: 1,
    totalCount: 1
  }