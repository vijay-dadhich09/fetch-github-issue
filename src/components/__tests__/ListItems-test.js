import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import ListItems from '../ListItems';
import { ISSUES_DATA } from '../../mock'

describe('Listitem test', () => {
  const props = ISSUES_DATA;
  it('renders correctly', () => {
    const tree = renderer
      .create(<ListItems {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders title', () => {
    const { getAllByA11yLabel, getByText } = render(<ListItems {...props} />);
    expect(getByText('List of Issues')).toHaveTextContent('List of Issues');
    expect(getByText('Automatic update of the `kubernetes` packages.')).toHaveTextContent('Automatic update of the `kubernetes` packages.');
    expect(getByText('Page 1 of 1')).toHaveTextContent('Page 1 of 1');
    expect(getByText('next')).toHaveTextContent('next');
    fireEvent.press(getByText('next'));
  });
});
