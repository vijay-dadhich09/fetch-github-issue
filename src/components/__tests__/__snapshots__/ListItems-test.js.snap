// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listitem test renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 10,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#273746",
        "fontSize": 20,
        "fontWeight": "800",
        "textAlign": "center",
      }
    }
  >
    List of Issues
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#C0392B",
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FDEDEC",
            "fontSize": 15,
            "fontWeight": "800",
            "textAlign": "center",
          }
        }
      >
        next
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#273746",
          "fontSize": 15,
          "fontWeight": "800",
          "textAlign": "center",
        }
      }
    >
      Page 1 of 1
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#C0392B",
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FDEDEC",
            "fontSize": 15,
            "fontWeight": "800",
            "textAlign": "center",
          }
        }
      >
        last
      </Text>
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    contentContainerStyle={
      Object {
        "marginTop": 10,
        "paddingBottom": 10,
      }
    }
    data={
      Array [
        Object {
          "id": 842912478,
          "title": "Automatic update of the \`kubernetes\` packages.",
        },
        Object {
          "id": 842668222,
          "number": 805,
          "title": "diskutils: add more robust handling of disk/partition operations",
        },
        Object {
          "id": 842369440,
          "title": "Change dropdown menus to accessible input fields",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#5DADE2",
              "justifyContent": "center",
              "minHeight": 60,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#09182b",
                "fontSize": 15,
                "fontWeight": "800",
              }
            }
          >
            Automatic update of the \`kubernetes\` packages.
          </Text>
        </View>
        <View
          style={
            Object {
              "height": 5,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#5DADE2",
              "justifyContent": "center",
              "minHeight": 60,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#09182b",
                "fontSize": 15,
                "fontWeight": "800",
              }
            }
          >
            diskutils: add more robust handling of disk/partition operations
          </Text>
        </View>
        <View
          style={
            Object {
              "height": 5,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#5DADE2",
              "justifyContent": "center",
              "minHeight": 60,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#09182b",
                "fontSize": 15,
                "fontWeight": "800",
              }
            }
          >
            Change dropdown menus to accessible input fields
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
